pyanaconda/modules/subscription/rhsm_observer.py

```
@@ -0,0 +1,73 @@
+SERVICE_ACTIVATION_TIMEOUT = 90.0
+
+class RHSMObserver(DBusObserver):
+    """Observer of the systemd unit backed RHSM DBus service."""
+
+    def __init__(self, message_bus, startup_check_method, timeout=SERVICE_ACTIVATION_TIMEOUT):
+        """Creates a RHSM service observer.
+
+        The observer will wait (up to a timeout) for the task starting the RHSM service
+        to finish and then provide access to the corresponding RHSM DBus API.
+
+        The waiting is only expected to happen at Anaconda startup, once the RHSM
+        startup task has finished running no further waiting is expected.
+
+        :param startup_check_method: a blocking method that waits for RHSM
+                                     startup to finish up to a given timeout
+                                     and then returns a boolean value corresponding
+                                     to True if startup was successful or False
+                                     otherwise
+        :param float timeout: how long to wait for RHSM service to start in seconds
+        """
+        super().__init__(message_bus, RHSM.service_name)
+        self._startup_check_method = startup_check_method
+        self._timeout = timeout
+
+    def get_proxy(self, object_identifier):
+        """Returns a proxy of the given RHSM object identifier.
+
+        :param object_identifier: RHSM DBus API object identifier
+        :type object_identifier: DBusObjectIdentifier instance
+        :raises: DBusObserverError if the given proxy can't be returned
+        """
+        can_return_proxy = False
+        # first check if the DBus API seems to be up
+        if self.is_service_available:
+            can_return_proxy = True
+        else:
+            log.debug("rhsm-observer: API not yet available, checking startup task state")
+            # the RHSM startup task is either still running or has already failed
+            can_return_proxy = self._startup_check_method(self._timeout)
```

You don't need the variable `can_return_proxy`. If `is_service_available` is False and `_startup_check_method` returns False, you can raise the exception. Otherwise, you can create the proxy.

---

Good point, let's do that.